portable/NetworkInterface/ThirdParty/MSP432/NetworkInterface.c

```diff
@@ -0,0 +1,981 @@
+    #undef ipconfigDRIVER_INCLUDED_TX_IP_CHECKSUM
+    #define ipconfigDRIVER_INCLUDED_TX_IP_CHECKSUM    1
+#endif
+#ifdef ipconfigDRIVER_INCLUDED_RX_IP_CHECKSUM
+    #undef ipconfigDRIVER_INCLUDED_RX_IP_CHECKSUM
+    #define ipconfigDRIVER_INCLUDED_RX_IP_CHECKSUM    1
+#endif
+#ifndef ipconfigETHERNET_MINIMUM_PACKET_BYTES
+    #define ipconfigETHERNET_MINIMUM_PACKET_BYTES     ETHERNET_MIN_PACKET_BYTES
+#endif
+#ifdef ipconfigETHERNET_MINIMUM_PACKET_BYTES
+    #undef ipconfigETHERNET_MINIMUM_PACKET_BYTES
+    #define ipconfigETHERNET_MINIMUM_PACKET_BYTES    ETHERNET_MIN_PACKET_BYTES
+#endif
+#if ipconfigHAS_DEBUG_PRINTF == 1
+    #warning The network interface may not work properly if ipconfigHAS_DEBUG_PRINTF 1.  Please set ipconfigHAS_DEBUG_PRINTF 0 for production use.
+#endif
+#ifndef __MSP432E401Y__
+    #define __MSP432E401Y__
+#endif
+
+static bool loadMACInternal();
+static bool setupEMAC();
+static void offEMAC();
+static void initDescriptors( uint32_t ul32Base );
+static uint32_t packetTransmit( uint8_t * pui8Buf,
+                                uint32_t ul32BufLen );
+static void ethernetIntHandler( void );
+static uint32_t processReceivedPacket( void );
+static void applicationProcessFrameRX( uint32_t ul32FrameLen,
+                                       uint8_t * uc8Buf,
+                                       uint32_t ulindex );
+static void prvCheckLinkUpOrDownTask( void * pvParameters );
+static void prvEMACDeferredInterruptHandlerTask( void * pvParameters );
+static bool isEMACLinkUp();
+static void DMAFreeDescriptorRX( uint32_t ulindx );
+static void prvEMACDeferredInterfaceOutput( void * pvParameters );
+static void prvChkNetworkState( void * pvParameters );
+static void vSetNetworkInterfaceConfig( const struct InternalNetworkInterfaceMSP432EConfig config );
+
+/* EMAC interrupts */
+#define EMAC_INTERRUPTS    EMAC_INT_RECEIVE | EMAC_INT_TRANSMIT | EMAC_INT_BUS_ERROR | EMAC_INT_TX_STOPPED | EMAC_INT_RX_STOPPED
+
+/* DMA descriptors */
+tEMACDMADescriptor g_psRxDescriptor[ NUM_TX_DESCRIPTORS ];
+tEMACDMADescriptor g_psTxDescriptor[ NUM_RX_DESCRIPTORS ];
+uint32_t ulg_ui32RxDescIndex;
+uint32_t ulg_ui32TxDescIndex;
+
+#define ETH_MAX_BUFFER_SIZE    1526
+#define ETH_RX_BUFFER_SIZE     ETH_MAX_BUFFER_SIZE
+#define ETH_TX_BUFFER_SIZE     ETH_MAX_BUFFER_SIZE
+uint8_t ucg_ppui8RxBuffer[ NUM_RX_DESCRIPTORS ][ ETH_RX_BUFFER_SIZE ];
+
+/* Task handles */
+TaskHandle_t xTaskToNotifyEthernetRX = NULL;
+TaskHandle_t xTaskToNotifyEthernetTX = NULL;
+TaskHandle_t xHandleCheckLinkUpOrDown = NULL;
+TaskHandle_t xHandleCheckNet = NULL;
+BaseType_t xHigherPriorityTaskWokenRX = pdFALSE;
+BaseType_t xHigherPriorityTaskWokenTX = pdFALSE;
+BaseType_t xHigherPriorityTaskWokenCheck = pdFALSE;
+
+/* Queue handles */
+QueueHandle_t xQueueRX;
+QueueHandle_t xQueueTX;
+
+/* MAC address */
+uint8_t ucpui8MACAddr[ ipMAC_ADDRESS_LENGTH_BYTES ];
+
+/* State variable that indicates whether the network is up or down */
+bool networkUP = false;
```

Comment:
> See my comments inline for specific items.
> 
> General comment – it might be helpful to list all of the tasks this network interface creates and their purpose in the documentation (I counted 5 tasks). Also, would it be possible to combine some of the tasks related to checking link up/down, network state & reset, etc.? Each task requires a stack, so if some of these could be combined it could help app developers save memory.

The link up/down, network state & reset tasks have now been combined as suggested.  As provided in the
README.md table, there are now only three tasks used in the driver code (NetworkInterface.c).

Reply:
networkUP is now static and external code can obtain the value of networkUP by calling vPublicCheckNetworkInterfaceUp().
The vPublicCheckNetworkInterfaceUp() is declared in the NetworkInterface.h file.